src/test/java/jssc/junit/rules/VirtualPortRule.java

```diff
@@ -0,0 +1,187 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class VirtualPortRule implements TestRule {
+
+  private static final Logger LOG = LoggerFactory.getLogger(VirtualPortRule.class);
+
+  private static final ExecutorService executor = Executors.newCachedThreadPool();
+
+  private final List<Future<?>> processes = new ArrayList<Future<?>>();
+
+  private OutputStream outputStream;
+
+  private boolean isAvailable;
+
+  private final Random random = new SecureRandom();
+
+  private File virtualCom1;
+  private File virtualCom2;
+
+  public VirtualPortRule() {
+  }
+
+  @Override
+  public Statement apply(final Statement base, final Description description) {
+    final String osName = System.getProperty("os.name", "").toLowerCase(Locale.ENGLISH);
+    if (!osName.startsWith("linux")) {
+      return new Statement() {
+        @Override
+        public void evaluate() throws Throwable {
+          base.evaluate();
+        }
+      };
+    }
+
+    // is linux.
+
+    initUnix(description);
+
+    return new Statement() {
+
+      @Override
+      public void evaluate() throws Throwable {
+        try {
+          final BackgroundProcess socatProcess = openSocat();
+          final Future<?> socatFuture = executor.submit(socatProcess);
+          VirtualPortRule.this.processes.add(socatFuture);
+
+          // Java 8:
+          // System.timemillis…
+          // while (!socatProcess.getProcess().isAlive()) {
+          //  Thread.sleep(50);
+          // }
+
+          // java 6 way… pretend it is running.
+          Thread.sleep(500);
+
+          VirtualPortRule.this.isAvailable = true;
+
+          base.evaluate();
+        } finally {
+          // stop socat
+          for (final Future<?> process : VirtualPortRule.this.processes) {
+            process.cancel(false);
+          }
+
+          executor.shutdown();
+        }
+      }
+    };
+  }
+
+  private void initUnix(final Description description) {
+    final File testTargetDir = new File("./target").getAbsoluteFile();
+    final File virtualPortsDir = new File(testTargetDir, "virtual-ports");
+
+    virtualPortsDir.mkdirs();
+
+    final String filename1 = String
+        .format("virtualcom-%s-%s-%d", description.getClassName(), description.getMethodName(), this.random.nextInt(10000));
+    this.virtualCom1 = new File(virtualPortsDir, filename1);
+    final String fileName2 = String
+        .format("virtualcom-%s-%s-%d", description.getClassName(), description.getMethodName(), this.random.nextInt(10000));
+    this.virtualCom2 = new File(virtualPortsDir, fileName2);
+  }
+
+  private BackgroundProcess openSocat() {
+    return new BackgroundProcess() {
+
+      private Process process;
+
+      @Override
+      public Process getProcess() {
+        return this.process;
+      }
+
+      @Override
+      public void run() {
+
+        try {
+          final ProcessBuilder processBuilder = new ProcessBuilder(asList(
+              "/usr/bin/socat",
```

Comment:
Wouldn't this be easier to just tackle with a scripting language like bash or python?

Reply:
Yes, I'd be happy for help here.

Reply:
I'm not sure how to write junit tests in scripting languages, so I'm resolving this conversation. 😅